README.md
# planet-rest-api [![Build status](https://travis-ci.org/twolfson/planet-rest-api.svg?branch=master)](https://travis-ci.org/twolfson/planet-rest-api)

REST API coding challenge for Planet Labs

## Requirements
Before getting started, make sure the following tools are installed locally:

- Python 2.7
- [virutalenvwrapper](https://virtualenvwrapper.readthedocs.io/en/latest/)

## Getting Started
To get a local copy of our repository running, run the following:

```bash
# Clone our repository
git clone git@github.com:twolfson/planet-rest-api.git
cd planet-rest-api

# Set up a virtual environment
mkvirtualenv planet-rest-api

# Install our dependencies
bin/bootstrap.sh

# Start our server
bin/start.sh
```

Our server should be running at <http://localhost:5000/>. For making requests, see our documentation.

**On future iterations, we can use:**

```bash
workon planet-rest-api
bin/start.sh
```

## Documentation
### Resources
We have the following resource types:

#### Asset
Dictionary containing information about physical asset. It has the following structure:

- name `String` - Name of asset
    - This must be unique across all access
    - This may only contain alphanumeric ASCII characters, underscores, and dashes
    - This cannot start with an underscore or dash
    - This must be between 4 and 64 characters long
- type `String` - Type of asset
    - This must be either "satellite" or "antenna"
- class `String` - Class of asset, depends on `type`
    - For a "satellite" asset, this must be either "dove" or "rapideye"
    - For an "antenna" asset, this must be either "dish" or "yagi"

**Examples:**

```json
{
  "name": "dove1",
  "type": "satellite",
  "class": "dove"
}
```

```json
{
  "name": "yagi3",
  "type": "antenna",
  "class": "yagi3"
}
```

### API
We expose a REST API with JSON-based resources with the following endpoints:

#### GET /
Basic health check for our API

**Request body:**

None

**Response body:**

- message `String` - OK response from server

**Example:**

```bash
$ curl http://localhost:5000/
{
  "message": "OK"
}
```

#### GET /assets
List all existing assets

**Request body:**

None

**Response body:**

- asset[] `list`
    - asset `Asset` - JSON representation of asset

**Example:**

```bash
$ curl http://localhost:5000/assets
[
  {
    "name": "dove1",
    "type": "satellite",
    "class": "dove"
  }
]
```

#### POST /assets
Create a new asset

**Request body:**

Parameters should match `Asset` specification (e.g. `name`, `type`)

**Response body:**

Upon success, response will match serialized `Asset` result

**Example:**

```bash
$ curl -X POST http://localhost:5000/assets \
  -H 'Content-Type: application/json' \
  --data '{
    "name": "dove1",
    "type": "satellite",
    "class": "dove"
  }'
{
  "name": "dove1",
  "type": "satellite",
  "class": "dove"
}
```

#### GET /assets/:name
Retrieve an asset by its name

**Request body:**

None

**Response body:**

Upon success, response will be JSON representation of asset

**Example:**

```bash
$ curl http://localhost:5000/assets/dove1
{
  "class": "dove",
  "name": "dove1",
  "type": "satellite"
}
```

#### Error handlers
If an error is encountered, then we will reply with the following format:

**Response body:**

- message `String` - Explanation of error

**Notes:**

We will also send an HTTP status code matching the error (e.g. 400, 404, 500).

### Testing
Tests can be run via:

```bash
# Runs `nosetests`
bin/test.sh
```

## Contributing
In lieu of a formal styleguide, take care to maintain the existing coding style. Add unit tests for any new or changed functionality. Test via `bin/test.sh`.

## Unlicense
As of Apr 10 2017, Todd Wolfson has released this repository and its contents to the public domain.

It has been released under the [UNLICENSE][].

[UNLICENSE]: UNLICENSE


CHANGELOG.md
# planet-rest-api changelog
1.0.0 - Initial release


planet_rest_api/test/test.py
# Load in our dependencies
from unittest import TestCase
import planet_rest_api


class TestRunFunction(TestCase):
    def test_run_exists(self):
        self.assertEqual(planet_rest_api.foo, 'bar')


planet_rest_api/test/__init__.py
# Load in our dependencies
# Taken from https://gist.github.com/twolfson/13f5f5784f67fd49b245
import json
from unittest import TestCase

import planet_rest_api
Asset = planet_rest_api.Asset


# Define our fixtures
# DEV: In a database setup, these would be separate files
#   but this is a trivial project
FIXTURES = {
    'satellite1': {
        'name': 'satellite1',
        'type': Asset.TYPE_SATELLITE,
        'klass': Asset.CLASS_SATELLITE_DOVE,
    }
}


# Define our base test case
class ApiTestCase(TestCase):
    # API setup/teardown
    @classmethod
    def setUpClass(cls):
        """On inherited classes, run our `setUp` method"""
        # Inspired via http://stackoverflow.com/questions/1323455/python-unit-test-with-base-and-sub-class/17696807#17696807 # noqa
        if cls is not ApiTestCase and cls.setUp is not ApiTestCase.setUp:
            orig_setUp = cls.setUp

            def setUpOverride(self, *args, **kwargs):
                ApiTestCase.setUp(self)
                return orig_setUp(self, *args, **kwargs)
            cls.setUp = setUpOverride

    def setUp(self):
        # Create our test client
        # http://flask.pocoo.org/docs/0.12/testing/
        planet_rest_api.app.config['TESTING'] = True
        self.app = planet_rest_api.app.test_client()

        # Wipe out our fixtures
        Asset._reset()

    def tearDown(self):
        # Reset our config
        del planet_rest_api.app.config['TESTING']

    # HTTP helper methods
    def _load_response(self, method, pathname, parse_json=True, *args, **kwargs):
        # Make our request
        method_fn = getattr(self.app, method)
        if 'data' in kwargs:
            kwargs['headers'] = kwargs.get('headers', {})
            kwargs['headers']['Content-Type'] = kwargs['headers'].get('Content-Type', 'application/json')
        rv = method_fn(pathname, *args, **kwargs)

        # Parse our JSON
        if parse_json:
            rv.json = json.loads(rv.data)

        # Return our result
        return rv

    def get_response(self, pathname, *args, **kwargs):
        return self._load_response('get', pathname, *args, **kwargs)

    def post_response(self, pathname, *args, **kwargs):
        return self._load_response('post', pathname, *args, **kwargs)

    # Define fixture helpers
    def install_fixtures(self, keys):
        # Resolve our fixtures
        # DEV: This will throw key errors when we can't find a fixture
        fixtures = [FIXTURES[key] for key in keys]

        # Install our fixtures
        for fixture in fixtures:
            asset = Asset(**fixture)
            asset.save()


planet_rest_api/test/root_get_test.py
# Load in our dependencies
from planet_rest_api.test import ApiTestCase


# Define our tests
class RootGetTestCase(ApiTestCase):
    def test_success(self):
        # A request to our root endpoint
        rv = self.get_response('/')

        # receives no errors
        self.assertEqual(rv.status_code, 200)
        self.assertEqual(rv.json, {'message': 'OK'})


planet_rest_api/__init__.py
# Load in our dependencies
from flask import Flask, jsonify, request
from flask_json_multidict import get_json_multidict

from planet_rest_api.models.asset import Asset


# Define our application
app = Flask(__name__)


# Install Multidict wrapper for JSON
# https://github.com/underdogio/flask-json-multidict/tree/1.0.0
# DEV: This is to get easy `400` errors with consistent format to urlencoded bodies
def resolve_request_body():
    """Before every request, resolve `request.body` from either `request.form` or `request.get_json()`"""
    request.body = request.form
    if request.headers.get('content-type') == 'application/json':
        request.body = get_json_multidict(request)
app.before_request(resolve_request_body) # noqa


# Bind our routes
@app.route('/', methods=['GET'])
def root():
    return jsonify({'message': 'OK'})


@app.route('/assets', methods=['GET'])
def assets_get():
    return jsonify([
        asset.serialize()
        for asset in Asset.get_all()
    ])


@app.route('/assets', methods=['POST'])
def assets_name_post():
    # Generate our asset
    # DEV: We are using `flask-json-multidict` which will raise a `BadRequestKeyError` on a missing parameter
    asset = Asset(
        name=request.body['name'],
        type=request.body['type'],
        klass=request.body['class'],
    )

    # Save and reply with our asset
    asset.save()
    return jsonify(asset.serialize())


@app.route('/assets/<name>', methods=['GET'])
def assets_name_get(name):
    asset = Asset.get_or_404(name)
    return jsonify(asset.serialize())


# Define our generic error handler
@app.errorhandler(400)
def invalid_request(error):
    # TODO: Add better messaging about which key is missing based on `BadRequestKeyError`
    return jsonify({'message': 'Invalid request'}), 400


@app.errorhandler(404)
def resource_not_found(error):
    return jsonify({'message': 'Resource not found'}), 404


planet_rest_api/models/asset.py
# Load in our dependencies
import re

from flask import abort


# Define our Asset class
class Asset(object):
    # Define our constants
    TYPE_SATELLITE = 'satellite'
    TYPE_ANTENNA = 'antenna'
    VALID_TYPES = [TYPE_SATELLITE, TYPE_ANTENNA]

    CLASS_SATELLITE_DOVE = 'dove'
    CLASS_SATELLITE_RAPIDEYE = 'rapideye'
    CLASS_ANTENNA_DISH = 'dish'
    CLASS_ANTENNA_YAGI = 'yagi'
    VALID_CLASSES = {
        TYPE_SATELLITE: [CLASS_SATELLITE_DOVE, CLASS_SATELLITE_RAPIDEYE],
        TYPE_ANTENNA: [CLASS_ANTENNA_DISH, CLASS_ANTENNA_YAGI],
    }

    # Define a store for all assets
    _asset_map = {}

    # Define our methods
    def __init__(self, name, type, klass):
        # Save our model information
        self.name = name
        self.type = type
        self.klass = klass

        # Mark our asset as new
        self.is_new = True

    @classmethod
    def _reset(cls):
        cls._asset_map = {}

    @classmethod
    def get_all(cls):
        return cls._asset_map.values()

    @classmethod
    def get(self, name):
        return self._asset_map.get(name)

    @classmethod
    def get_or_404(self, name):
        if name not in self._asset_map:
            abort(404)
        else:
            return self._asset_map[name]

    def save(self):
        # Run validation
        self.validate()

        # If our asset is new
        if self.is_new is True:
            # Assert it isn't overwriting an asset
            if self.name in self._asset_map:
                abort(400)

            # Save it
            self._asset_map[self.name] = self

        # Otherwise, do nothing -- it's already in our db
        self.is_new = False

    def validate(self):
        # Validate our asset name
        # DEV: We could use a custom error here and add a custom handler
        #   but for this exercise, a Flask error should work
        # DEV: We assert length first to avoid errors with `self.name[0]`
        if len(self.name) < 4:
            abort(400, 'Name is too short (under 4 characters)')
        if len(self.name) > 64:
            abort(400, 'Name is too long (over 64 characters)')
        if re.search(r'[^A-Za-z0-9\-_]', self.name):
            abort(400, 'Name contains non-alphanumeric, dash, or underscore characters')
        if self.name[0] in ['-', '_']:
            abort(400, 'Name begins with a dash or underscore')

        # Validate our asset type
        if self.type not in self.VALID_TYPES:
            abort(400, 'Invalid asset type. Should be "satellite" or "antenna"')

        # Verify our type is in our matching classes as sanity for developers
        if self.type not in self.VALID_CLASSES:
            raise RuntimeError('Encountered unexpected asset type "{type}"'.format(type=self.type))

        # Validate our asset class matches our type
        matching_classes = self.VALID_CLASSES[self.type]
        if self.klass not in matching_classes:
            abort(400, 'Invalid asset class. Should match asset\'s expected types')

    def serialize(self):
        return {
            'name': self.name,
            'type': self.type,
            'class': self.klass,
        }
